woltka/biom.py

```diff
@@ -43,8 +43,8 @@ def profile_to_biom(profile, samples=None, tree=None, rankdic=None,
     biom.Table
         Converted BIOM table.
     """
-    # features = sorted(allkeys(profile))
-    samples = samples or sorted(profile)
+    samples = [x for x in samples if x in profile] if samples else sorted(
+        profile)
```

Comment:
@fedarko Good advice! Here the purpose of `samples` is to allow user to provide an **ordered** list of sample IDs to be shown in the table header. Therefore order matters. I modified the code to make it more readable but let me not adopt the set intersection solution because it will eliminate the order of elements.

Meanwhile, both `samples` and `profile` won't be large. Their elements are sample IDs, and I guess a study won't have more than a few thousand samples. Plus, this function only needs to be executed once in the end of the workflow to generate an output BIOM table. Therefore, comparing to many other steps, I think this step does not need to be rigorously optimized.

And finally, there is unit test, in `test_biom.py`. Thanks for noting this. I added another test to explicit test the scenario where some sample IDs do not exist in the profile.

I also modified `write_table` below accordingly.

Reply:
Got it! That all sounds good to me -- thanks for the detailed and thoughtful response :)